### PR TITLE
Fixes for MobileBERT pruning and added tests for pruning transformations

### PR DESCRIPTION
**Pull request by ljaljushkin**
### Changes

Filter out groups that can't be matched with the given unstructured pruning structures.
Supported pruning by columns and rows at the same time in the JPQD.
Added tests for pruning transformations for BERT, CLIPVision, Swin, MobileBERT, Wave2vec2, DistilBert

### Reason for changes

Automatically found structured pruning groups for MobileBERT and all others models should be properly matched with the unstructured pruning structures defined by the user in the config.

Tests for ngraph pruning transformation should find issues with pruning groups on early stages without long cycle of validation in the OV. 3 of them revealing new issues in the transformation for 3 models: MobileBERT, DistilBERT and Wave2vec2. 

### Related tickets

108495

### Tests

test_ngraph_pruning
test_populate_dependent_structured_mask_by_row_and_column

### CODE REVIEW

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py [hunk 1]
```diff
@@ -169,14 +169,31 @@ def update_independent_structured_mask_from_operand(self):
...
+        if self.sparsifier_operand.prune_bias:
+            self.sparsifier_operand.bias_ctx.binary_mask.fill_(1)
+
     def populate_dependent_structured_mask_to_operand(self):
         """
         Updates the actual binary masks in operand with `self.dependent_structured_mask`.
         """
         structured_mask_inflated = self._inflate_structured_mask(self.dependent_structured_mask, self.grid_size)
-        self.sparsifier_operand.weight_ctx.binary_mask = structured_mask_inflated
+        self.sparsifier_operand.weight_ctx.binary_mask *= structured_mask_inflated
```
Comment (ljaljushkin) [inline nncf/experimental/torch/sparsity/movement/structured_mask_handler.py:194]: @vuiseng9 @yujiepan-work 
Assume that controller initializes binary mask by ones. If some linear layer is pruned by row and column at the same time, the corresponding sparsifier operand will update binary 2 times: first multiply ones with the mask per row, second – multiply the produced mask with per column. I've verified it with 2 tests: 
[test_populate_dependent_structured_mask_by_row_and_column](https://github.com/openvinotoolkit/nncf/pull/1724/files#diff-2c85b01343c47e2fab4bcc0c8b0f78549343b2ff261cdc7c124660d790cce0d2R233)
[the ngraph transformation test for MobileBERT's recipe with all layers sparsified by default in fine-grained mode](https://github.com/openvinotoolkit/nncf/pull/1724/files#diff-65c84e2828d4233ff8ba02f754b7a5a644a922006cae2ad0346529e96ae11934R189)

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py [hunk 2]
```diff
@@ -320,6 +337,9 @@ def populate_dependent_structured_mask_to_operand(self):
         """
         Asks all contexts in `self._structured_mask_ctx_groups` to update the actual binary masks in operand.
         """
+        for group in self._structured_mask_ctx_groups:
+            for ctx in group.structured_mask_contexts:
+                ctx.initialize_binary_mask()
```
Comment (ljaljushkin) [inline nncf/experimental/torch/sparsity/movement/structured_mask_handler.py:342]: @vuiseng9 @yujiepan-work At this point unstructured binary mask was overridden anyway. I just initialize it by ones first to apply inflated structured masks (per row or/and per column) by multiply (or logic AND).

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py [hunk 3]
```diff
@@ -381,24 +401,35 @@ def _create_structured_mask_context_groups(
...
             for producer_info in extended_producers:
                 nncf_node = nncf_graph.get_node_by_id(producer_info.node_id)
                 module = nncf_network.nncf.get_containing_module(nncf_node.node_name)
                 if module in module_vs_sparse_module_info_map:
                     # 0 dimension corresponds to row (output channels), 1st dimension - to column (input channels)
                     prune_by_row = not bool(producer_info.pruning_dimension)
                     minfo = module_vs_sparse_module_info_map[module]
+                    sparsifier: MovementSparsifier = minfo.operand
+                    if sparsifier.sparse_structure == SparseStructure.PER_DIM:
+                        sparse_axis = sparsifier.sparse_cfg.sparse_axis
```
Comment (ljaljushkin) [inline nncf/experimental/torch/sparsity/movement/structured_mask_handler.py:416]: Filter out groups that have structured pruning per dimension that doesn't match the given dimension for unstructured pruning

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py [hunk 4]
```diff
@@ -381,24 +401,35 @@ def _create_structured_mask_context_groups(
...
                                                 prune_by_row)
                     ctxes.append(ctx)
                 else:
-                    nncf_logger.warning(f'Automatically found structured pruning group does not match the given '
-                                        f'unstructured sparse structures:\n {group}')
+                    nncf_logger.warning(f'{prefix_warning}. No unstructured pruning for: {nncf_node.node_name}')
+                    is_group_matched = False
                     break
-            if ctxes:
+            if ctxes and is_group_matched:
```
Comment (ljaljushkin) [inline nncf/experimental/torch/sparsity/movement/structured_mask_handler.py:433]: Don't add group if there's any unmatched producer

File: tests/torch/sparsity/movement/test_model_saving.py [hunk 5]
```diff
@@ -109,6 +121,145 @@ def test_same_outputs_in_torch_and_exported_onnx(self, tmp_path: Path, recipe: B
         assert np.allclose(softmax(onnx_outputs, axis=-1),
                            softmax(torch_outputs, axis=-1), atol=1e-6)
 
+    @pytest.mark.skipif(parse_version(torch.__version__) < parse_version("1.12"),
+                    reason=f"torch {torch.__version__} is not compatible with installed transformers package. "
+                            f"Some tests may fail with segmentation fault")
+    @pytest.mark.parametrize('desc', [
+        Dict(
+            nncf_weight_ratio=0.43,
```
Comment (ljaljushkin) [inline tests/torch/sparsity/movement/test_model_saving.py:129]: Removing 75% channels in the MHA and FFN doesn't give the same sparsity rate for the whole model, since dimensions are tiny and embeddings are not pruned.

File: tests/torch/sparsity/movement/test_model_saving.py [hunk 6]
```diff
@@ -109,6 +121,145 @@ def test_same_outputs_in_torch_and_exported_onnx(self, tmp_path: Path, recipe: B
...
+            ov_weight_ratio=0.36,
```
Comment (ljaljushkin) [inline tests/torch/sparsity/movement/test_model_saving.py:130]: There are some constants in the OpenVINO that gives slightly different compression rate by number of parameters in the IR. Manually verified graphs and dimensions and empirically found that 15% is the maximum acceptable difference in comparison with rate calculated in NNCF. In that case, test assumed to be passed.

File: tests/torch/sparsity/movement/test_model_saving.py [hunk 7]
```diff
@@ -109,6 +121,145 @@ def test_same_outputs_in_torch_and_exported_onnx(self, tmp_path: Path, recipe: B
...
+            recipe=SwinRunRecipe().model_config_(
+                num_heads=[4],
+                num_labels=1,
+                num_channels=4,
+                window_size=1,
+            ),
+        ),
+        Dict(
+            nncf_weight_ratio=0.55,
+            ov_weight_ratio=0.20,
```
Comment (ljaljushkin) [inline tests/torch/sparsity/movement/test_model_saving.py:183]: Big difference for DistilBERT reveals issue with attention mask.
Broadcast and select is not properly supported in the NGraph transformation.
MicrosoftTeams-image (5)

File: tests/torch/sparsity/movement/test_model_saving.py [hunk 8]
```diff
@@ -109,6 +121,145 @@ def test_same_outputs_in_torch_and_exported_onnx(self, tmp_path: Path, recipe: B
...
+            recipe=DistilBertRunRecipe().model_config_(),
+        ),
+        Dict(
+            nncf_weight_ratio=0.63,
+            ov_weight_ratio=0.38,
+            recipe=MobileBertRunRecipe().model_config_(),
```
Comment (ljaljushkin) [inline tests/torch/sparsity/movement/test_model_saving.py:189]: Big difference for MobileBERT reveals some bug with pruning by row and column at the same time.
MicrosoftTeams-image (6)

File: tests/torch/sparsity/movement/test_model_saving.py [hunk 9]
```diff
@@ -109,6 +121,145 @@ def test_same_outputs_in_torch_and_exported_onnx(self, tmp_path: Path, recipe: B
...
+                intermediate_size=4,
+                num_labels=1,
+                num_conv_pos_embeddings=1,
+                num_attention_heads=4,
+                classifier_proj_size=1
+            )
+        ),
+        Dict(
+            nncf_weight_ratio=0.41,
+            ov_weight_ratio=0.07,
```
Comment (ljaljushkin) [inline tests/torch/sparsity/movement/test_model_saving.py:154]: Will open the issue for Wave2vec2
can't prune the path with the following reshape: 
MicrosoftTeams-image (4)

**Review by daniil-lyakhov — APPROVED** [review]
Looks ok from node search algo perspective 👍

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py [hunk 10]
```diff
@@ -27,7 +27,7 @@
 from nncf.torch.layers import NNCFLinear
 from nncf.torch.nncf_network import NNCFNetwork
 from nncf.torch.sparsity.base_algo import SparseModuleInfo
-from nncf.experimental.torch.sparsity.movement.layers import MovementSparsifier
+from nncf.experimental.torch.sparsity.movement.layers import MovementSparsifier, SparseStructure
```
Comment (alexsu52) [inline nncf/experimental/torch/sparsity/movement/structured_mask_handler.py:30]:
```suggestion
from nncf.experimental.torch.sparsity.movement.layers import MovementSparsifier
from nncf.experimental.torch.sparsity.movement.layers import SparseStructure
```

File: tests/torch/requirements.txt [hunk 11]
```diff
@@ -14,4 +14,5 @@ transformers==4.23.1
 # Required for movement_sparsity tests
 datasets==2.6.2
 evaluate==0.3.0
-timm
\ No newline at end of file
+timm==0.6.13
+openvino-dev==2023.0.0.dev20230217
```
Comment (alexsu52) [inline tests/torch/requirements.txt:18]: openvino-dev~=2023.0.0.dev

The last dev or release openvino-dev package will be installed.

**Review by alexsu52 — COMMENTED** [review]
LGTM

File: tests/torch/requirements.txt (hunk 11 above)
Comment (ljaljushkin) [reply tests/torch/requirements.txt]: thanks, corrected

File: nncf/experimental/torch/sparsity/movement/structured_mask_handler.py (hunk 10 above)
Comment (ljaljushkin) [reply nncf/experimental/torch/sparsity/movement/structured_mask_handler.py]: Now it's resolved automatically

**Review by alexsu52 — APPROVED** [review]
LGTM
